### ISSUE TITLE
Unable to run ansible commands

### ISSUE THREAD

**Issue by orderedchaosdev** [ansible/ansible#12257]
Issue Type:

Bug Report

Ansible Version:

ansible 2.0.0 (devel 70677b030b) last updated 2015/09/04 16:32:33 (GMT -700)
  lib/ansible/modules/core: (detached HEAD 2520627fe7) last updated 2015/09/04 16:32:38 (GMT -700)
  lib/ansible/modules/extras: (detached HEAD 5cd0fab9d0) last updated 2015/09/04 16:32:42 (GMT -700)

Ansible Configuration:

transport=ssh

Environment:

Centos 6.7

Summary:

Unable to run any ansible commands

Steps To Reproduce:

ansible host1 -m ping

Expected Results:

```
ansible host1 -m ping
host1 | SUCCESS => {
    "changed": false, 
    "ping": "pong"
}
```

Actual Results:

```
ansible host1 -m ping
ERROR! compile() expected string without null bytes
```

As of Commit 87f75a5 it worked but not
with
Commit ff9f5d7 or devel

**Comment by jimi-c** [comment]
@orderedchaosdev I'm unable to reproduce this. Can you test with the most recent pull from devel and see if you're still having the problem?

**Comment by orderedchaosdev** [comment]
I am still having the issue if I change the python path to the older version of ansible then it works. I keep backups when I test updates

**Comment by orderedchaosdev** [comment]
strace on it makes it look like it is occurring while it is parsing the inventory file 

```
getcwd("/opt/servermgr/tests", 1024)    = 21
stat("/opt/servermgr/tests/prod", {st_mode=S_IFREG|0600, st_size=45, ...}) = 0
getcwd("/opt/servermgr/tests", 1024)    = 21
stat("/opt/servermgr/tests/prod", {st_mode=S_IFREG|0600, st_size=45, ...}) = 0
open("prod", O_RDONLY)                  = 3
fstat(3, {st_mode=S_IFREG|0600, st_size=45, ...}) = 0
fstat(3, {st_mode=S_IFREG|0600, st_size=45, ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f8ad98c6000
read(3, "[lab]\nhost-1 ansible_ssh_host=16"..., 8192) = 45
read(3, "", 4096)                       = 0
close(3)                                = 0
munmap(0x7f8ad98c6000, 4096)            = 0
getcwd("/opt/servermgr/tests", 1024)    = 21
stat("/opt/servermgr/tests/prod", {st_mode=S_IFREG|0600, st_size=45, ...}) = 0
stat("/opt/servermgr/tests/prod", {st_mode=S_IFREG|0600, st_size=45, ...}) = 0
stat("/opt/servermgr/tests/prod", {st_mode=S_IFREG|0600, st_size=45, ...}) = 0
getcwd("/opt/servermgr/tests", 1024)    = 21
stat("/opt/servermgr/tests/prod", {st_mode=S_IFREG|0600, st_size=45, ...}) = 0
getcwd("/opt/servermgr/tests", 1024)    = 21
stat("/opt/servermgr/tests/prod", {st_mode=S_IFREG|0600, st_size=45, ...}) = 0
open("prod", O_RDONLY)                  = 3
fstat(3, {st_mode=S_IFREG|0600, st_size=45, ...}) = 0
fstat(3, {st_mode=S_IFREG|0600, st_size=45, ...}) = 0
lseek(3, 0, SEEK_CUR)                   = 0
fstat(3, {st_mode=S_IFREG|0600, st_size=45, ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f8ad98c6000
lseek(3, 0, SEEK_CUR)                   = 0
read(3, "[lab]\nhost-1 ansible_ssh_host=16"..., 4096) = 45
read(3, "", 4096)                       = 0
close(3)                                = 0
munmap(0x7f8ad98c6000, 4096)            = 0
write(2, "\33[0;31mERROR! compile() expected"..., 62ERROR! compile() expected string without null bytes) = 62
write(2, "\n", 1
```

**Comment by amenonsen** [comment]
@orderedchaosdev any chance you could provide a minimal inventory that reproduces this problem?

**Comment by orderedchaosdev** [comment]
```
[lab]
host-1 ansible_ssh_host=192.168.1.1
```

IP address is not the actual of course.

config

```
[defaults]
forks=50
log_path= ~/ansible-logs/ansible.log
library= ~/<ansible checkout path>/libary:~/ansible-library
remote_user=root
transport=ssh
roles_path=<role path>
[ssh_connection]
ssh_args=-o StrictHostKeyChecking=true
```

This is an example config
I defined the inventory file in the command line for the test

**Comment by amenonsen** [comment]
Well, that works fine for me, and I can't figure out how any of the `re.compile()` calls in `inventory/ini.py` could get null bytes in the input. I don't know how to approach this problem.

**Comment by orderedchaosdev** [comment]
I modified the inventory file to remove the ansible_ssh_host part and then I got this message instead

```
-> ansible-playbook lookup.yml -i prod -vvv
Using /opt/servermgr/ansible-playbooks/ansible_testing.cfg as config file
Unexpected Exception: lstat() argument 1 must be encoded string without NULL bytes, not str
the full traceback was:

Traceback (most recent call last):
  File "<ansible_path>/bin/ansible-playbook", line 79, in <module>
    sys.exit(cli.run())
  File "<ansible_path>/lib/ansible/cli/playbook.py", line 116, in run
    inventory = Inventory(loader=loader, variable_manager=variable_manager, host_list=self.options.inventory)
  File "<ansible_path>/lib/ansible/inventory/__init__.py", line 77, in __init__
    self.parse_inventory(host_list)
  File "<ansible_path>/lib/ansible/inventory/__init__.py", line 133, in parse_inventory
    host.vars = combine_vars(host.vars, self.get_host_variables(host.name))
  File "<ansible_path>/lib/ansible/inventory/__init__.py", line 499, in get_host_variables
    self._vars_per_host[hostname] = self._get_host_variables(hostname, vault_password=vault_password)
  File "<ansible_path>/lib/ansible/inventory/__init__.py", line 529, in _get_host_variables
    vars = combine_vars(vars, self.get_host_vars(host))
  File "<ansible_path>/lib/ansible/inventory/__init__.py", line 653, in get_host_vars
    return self._get_hostgroup_vars(host=host, group=None, new_pb_basedir=new_pb_basedir)
  File "<ansible_path>/lib/ansible/inventory/__init__.py", line 702, in _get_hostgroup_vars
    base_path = os.path.realpath(os.path.join(basedir, "host_vars/%s" % host.name))
  File "/usr/lib64/python2.6/posixpath.py", line 365, in realpath
    if islink(component):
  File "/usr/lib64/python2.6/posixpath.py", line 132, in islink
    st = os.lstat(path)
TypeError: lstat() argument 1 must be encoded string without NULL bytes, not str

```

**Comment by orderedchaosdev** [comment]
Also I modified the inventory file to remove the ansible_ssh_host part and then I got this message instead

**Comment by orderedchaosdev** [comment]
If you can think of additional tests I can try them. And I just pulled the current devel and the issue is still present

**Comment by nitzmahone** [comment]
@amenonsen - I just got nailed by this on a test box as well; It's due to the addition of shlex + unicode in the INI inventory parser. shlex.split() was horribly broken on unicode input from Python 2.5.2 to 2.7.2 (see https://bugs.python.org/issue6988). Perhaps use one of the ushlex impls, or do the split a different way?

**Comment by amenonsen** [comment]
@nitzmahone do you know if using `to_bytes` before and `to_unicode` after (as @abadger suggested to me on IRC) would work? I can make the change (it would be pretty trivial), but I am not in a position to test.

BTW, given the next stack trace that @orderedchaosdev posted above, it sounds like there are bigger non-inventory-parsing-related problems with NUL bytes.

**Comment by abadger** [comment]
@orderedchaosdev @nitzmahone Are either of you able to provide your inventory?  Perhaps privately if you don't want it posted to a public bug?

**Comment by orderedchaosdev** [comment]
I posted my sample inventory that was all of it in the test

**Comment by abadger** [comment]
Unfortunatey the sample inventory isn't reproducing the bug here.  Is just that sample inventory enough to reproduce the bug for you?  If so, perhaps you could upload that sample inventory as a file somewhere and we could download that?

**Comment by nitzmahone** [comment]
Any inventory should show the problem on a broken Python version. Try a CentOS 6.6 instance (ships w/ Python 2.6.6). I broke it with:
`localhost ansible_connection=local`

**Comment by nitzmahone** [comment]
@abadger @amenonsen I wouldn't round-trip the inventory through to_bytes/to_unicode- interpreting deconstructed UCS2/4/UTF-8 as ASCII is a recipe for trouble when there are actually multi-byte code points/surrogates in the text (presumably the reason we're doing Unicode strings in the first place). The potential exists for a large number of code points to trigger weird behavior in the parser. If the inventory parser's going to be Unicode-aware, everything that touches the data needs to be Unicode-aware (which means a ushlex.split() or a more manual approach for the ini parsing).

**Comment by abadger** [comment]
@nitzmahone ah that's right.. I forgot that shlex is only broken in python2.6 and a few versions of python2.5 and python2.7.

I hear what you're  saying about not trusting a parser designed for bytes to operate correctly on non-asci values but there aren't many good options.  Using a third-party library to parse would introduce a dependency which ansible probably wants to avoid.  We're already maintaining a few parsers and for 2.0 have been trying to get rid of them rather than code more.  On the plus side for shlex,  jimi-c has been using shlex with data translated to str for a while with success.  shlex does try to handle non-ascii data.  It just doesn't make any promises of working with the unicode type (as opposed to encoded strings) until 2.7.3.

Looking at the shlex code, almost all of the tokens its looking for are ascii compatible.  Since ascii is a subset of utf8 and the ascii byte values are unique in utf-8 those shouldn't be a source of problems problems.  The only one that I have misgivings about is wordchars.  It's documented as "The string of characters that will accumulate into multi-character tokens. By default, includes all ASCII alphanumerics and underscore."  That troubles me as it seems like we'd want non-ascii characters to also accumulate into multi-character tokens.  However, testing shlex.split() seems to indicate that non-ascii characters are being accumulated despite not being present in wordchars.  So either I don't quite understand what the documentation means or there's some other mechanism in addition to wordchars that is at work here.

As for why we're using unicode strings, we're trying to use unicode strings internally for everything.  I do think that we want to allow non-ascii in the inventory files but the basic reason we're processing unicode type instead of byte type here is that the data has been read in from its source and then turned into unicode so that str type doesn't escape into the rest of the code.

So thinking about all of this, I think the chance of bugs is probably higher if we write our own code.  And since we can't use a third party shlex implementation, using the stdlib's shlex with byte strs is probably the best choice until we run up against a limitation of it that actually causes us problems.

**Comment by abadger** [comment]
@orderedchaosdev , that commit solved the issue for me on centos6.  Feel free to let me know if the problem is still there for you.

@nitzmahone If you can think of a concrete way that shlex is still going to break things, let me know and I'll discuss what to do about it with jimi-c .

# Closing This Ticket

Hi!

We believe recent commits (likely detailed above) should resolve this question or problem for you.

This will also be included in the next major release.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular
issue is resolved.

Thank you!

**Comment by towertop** [comment]
Which version has the fix? I got it again in 2.0.2 when installing it with pip.

I ran into a similar failure on 2.0.2, whenever I was trying to use variables in inventory file.  If I just use a hostname in inventory, it would be ok. I guess that is related to the blank space as separators.

```
~ # ansible blog -vvv -m ping
Using /root/.ansible.cfg as config file
ERROR! Unexpected Exception: must be encoded string without NULL bytes, not str
the full traceback was:

Traceback (most recent call last):
  File "/usr/bin/ansible", line 86, in <module>
    sys.exit(cli.run())
  File "/usr/lib/python2.7/site-packages/ansible/cli/adhoc.py", line 126, in run
    inventory = Inventory(loader=loader, variable_manager=variable_manager, host_list=self.options.inventory)
  File "/usr/lib/python2.7/site-packages/ansible/inventory/__init__.py", line 85, in __init__
    self.parse_inventory(host_list)
  File "/usr/lib/python2.7/site-packages/ansible/inventory/__init__.py", line 147, in parse_inventory
    host.vars = combine_vars(host.vars, self.get_host_variables(host.name))
  File "/usr/lib/python2.7/site-packages/ansible/inventory/__init__.py", line 540, in get_host_variables
    self._vars_per_host[hostname] = self._get_host_variables(hostname, vault_password=vault_password)
  File "/usr/lib/python2.7/site-packages/ansible/inventory/__init__.py", line 570, in _get_host_variables
    vars = combine_vars(vars, self.get_host_vars(host))
  File "/usr/lib/python2.7/site-packages/ansible/inventory/__init__.py", line 702, in get_host_vars
    return self._get_hostgroup_vars(host=host, group=None, new_pb_basedir=new_pb_basedir)
  File "/usr/lib/python2.7/site-packages/ansible/inventory/__init__.py", line 749, in _get_hostgroup_vars
    results = combine_vars(results, self._variable_manager.add_host_vars_file(base_path, self._loader))
  File "/usr/lib/python2.7/site-packages/ansible/vars/__init__.py", line 545, in add_host_vars_file
    (name, data) = self._load_inventory_file(path, loader)
  File "/usr/lib/python2.7/site-packages/ansible/vars/__init__.py", line 503, in _load_inventory_file
    if loader.is_directory(path):
  File "/usr/lib/python2.7/site-packages/ansible/parsing/dataloader.py", line 131, in is_directory
    return os.path.isdir(to_bytes(path, errors='strict'))
  File "/usr/lib/python2.7/genericpath.py", line 49, in isdir
    st = os.stat(s)
TypeError: must be encoded string without NULL bytes, not str
```

My versions:

```
~ # pip list -v
ansible (2.0.2.0)
ecdsa (0.13)
Jinja2 (2.8)
MarkupSafe (0.23)
paramiko (1.16.0)
pip (8.1.1)
pycrypto (2.6.1)
PyYAML (3.11)
setuptools (20.9.0)

~ # python --version
Python 2.7.11
```

The ~/ansible_hosts likes:

```
blog ansible_host=myhost.cn ansible_user=root ansible_ssh_private_key_file=/root/.ssh/my_rsa
```
